### DISCUSSION TITLE
Shall we merge hnix and it's store ? (and some more questions :-))

### DISCUSSION THREAD

**Discussion by layus** [haskell-nix/hnix#827] [Show and tell]
I got some real speedup (25% faster) by... using a readonly store :tada: 

```
$ hyperfine --warmup 1 --min-runs 3 blablabla
Benchmark #1: readonly (fake) store
  Time (mean ± σ):      8.719 s ±  0.019 s    [User: 8.405 s, System: 0.313 s]
  Range (min … max):    8.704 s …  8.741 s    3 runs
 
Benchmark #2: remote (daemon) store
  Time (mean ± σ):     11.495 s ±  0.024 s    [User: 8.194 s, System: 0.464 s]
  Range (min … max):   11.468 s … 11.513 s    3 runs
 
Summary
  'readlonly store' ran
    1.32 ± 0.00 times faster than 'remote store'
```
Not really a surprise, and yet there is something to investigate because nix does not pay this 3s penalty for using a remote store. Did I miss some optimisation somewhere ?

---

Anyway, the point of this discussion is to come with a real example where changes in hnix and hnix-store are (again) synchonized:

```diff
diff --git a/hnix/src/Nix/Effects.hs b/hnix/src/Nix/Effects.hs
index 0ce82e6..76c6446 100644
--- a/hnix/src/Nix/Effects.hs
+++ b/hnix/src/Nix/Effects.hs
@@ -43,6 +43,7 @@ import qualified System.Nix.Hash               as Store
 import qualified System.Nix.Store.Remote       as Store
 import qualified System.Nix.Store.Remote.Types as Store
 import qualified System.Nix.StorePath          as Store
+import qualified System.Nix.ReadonlyStore      as Store
 
 -- | A path into the nix store
 newtype StorePath = StorePath { unStorePath :: FilePath }
@@ -262,22 +263,21 @@ parseStoreResult name res = case res of
   (Left msg, logs) -> return $ Left $ ErrorCall $ "Failed to execute '" ++ name ++ "': " ++ msg ++ "\n" ++ show logs
   (Right result, _) -> return $ Right result
 
+toStorePath :: Store.StorePath -> StorePath
+toStorePath = StorePath . T.unpack . T.decodeUtf8 . Store.storePathToRawFilePath
+
 instance MonadStore IO where
 
   addToStore name path recursive repair = case Store.makeStorePathName name of
     Left err -> return $ Left $ ErrorCall $ "String '" ++ show name ++ "' is not a valid path name: " ++ err
     Right pathName -> do
       -- TODO: redesign the filter parameter
-      res <- Store.runStore $ Store.addToStore @'Store.SHA256 pathName path recursive (const False) repair
-      parseStoreResult "addToStore" res >>= \case
-        Left err -> return $ Left err
-        Right storePath -> return $ Right $ StorePath $ T.unpack $ T.decodeUtf8 $ Store.storePathToRawFilePath storePath
-
-  addTextToStore' name text references repair = do
-    res <- Store.runStore $ Store.addTextToStore name text references repair
-    parseStoreResult "addTextToStore" res >>= \case
-      Left err -> return $ Left err
-      Right path -> return $ Right $ StorePath $ T.unpack $ T.decodeUtf8 $ Store.storePathToRawFilePath path
+      Right . toStorePath <$> Store.computeStorePathForPath @'Store.SHA256 pathName path recursive (const False) repair
+
+  addTextToStore' name text references _repair = case Store.makeStorePathName name of
+    Left err -> return $ Left $ ErrorCall $ "String '" ++ show name ++ "' is not a valid path name: " ++ err
+    Right pathName -> do
+      return $ Right $ toStorePath $ Store.computeStorePathForText "/nix/store" pathName (T.encodeUtf8 text) references
 
 addTextToStore :: (Framed e m, MonadStore m) => StorePathName -> Text -> Store.StorePathSet -> RepairFlag -> m StorePath
 addTextToStore a b c d = either throwError return =<< addTextToStore' a b c d
diff --git a/hnix-store/hnix-store-core/src/System/Nix/ReadonlyStore.hs b/hnix-store/hnix-store-core/src/System/Nix/ReadonlyStore.hs
index 84e37e4..4d77240 100644
--- a/hnix-store/hnix-store-core/src/System/Nix/ReadonlyStore.hs
+++ b/hnix-store/hnix-store-core/src/System/Nix/ReadonlyStore.hs
@@ -2,6 +2,9 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+
 module System.Nix.ReadonlyStore where
 
 import           Data.ByteString (ByteString)
@@ -10,7 +13,9 @@ import qualified Data.Text as T
 import qualified Data.HashSet as HS
 import           Data.Text.Encoding
 import           System.Nix.Hash
+import           System.Nix.Nar
 import           System.Nix.StorePath
+import           Control.Monad.State.Strict
 
 makeStorePath :: forall hashAlgo . (NamedAlgo hashAlgo)
   => FilePath
@@ -49,3 +54,22 @@ makeFixedOutputPath fp recursive h nm =
 
 computeStorePathForText :: FilePath -> StorePathName -> ByteString -> StorePathSet -> StorePath
 computeStorePathForText fp nm s refs = makeTextPath fp nm (hash s) refs
+
+computeStorePathForPath :: forall a. (ValidAlgo a, NamedAlgo a)
+           => StorePathName        -- ^ Name part of the newly created `StorePath`
+           -> FilePath             -- ^ Local `FilePath` to add
+           -> Bool                 -- ^ Add target directory recursively
+           -> (FilePath -> Bool)   -- ^ Path filter function
+           -> Bool                 -- ^ Only used by local store backend
+           -> IO StorePath
+computeStorePathForPath name pth recursive _pathFilter _repair = do
+  selectedHash <- if recursive then recursiveContentHash else flatContentHash
+  pure $ makeFixedOutputPath "/nix/store" recursive selectedHash name
+ where
+  recursiveContentHash :: IO (Digest a)
+  recursiveContentHash = finalize @a <$> execStateT streamNarUpdate (initialize @a)
+  streamNarUpdate :: StateT (AlgoCtx a) IO ()
+  streamNarUpdate = streamNarIO (\t -> modify $ flip (update @a) t) narEffectsIO pth
+
+  flatContentHash :: IO (Digest a)
+  flatContentHash = hashLazy <$> narReadFile narEffectsIO pth
```

--- 

Also, this implementation only works because we do some hash caching in hnix, which we added to speed-up the remote store implem. But with this caching we do not have to use the remote store.

I think a readonly, in-memeory store is a saner default for hnix, especially for tests. Or at least a --readonly option for the tests.

---

And my last problem is: How to switch dynamically the MonadStore instance based on the Options we have (--readonly, --store, etc ?). From what I understand from haskell, type classes may not be the right tool.

**Comment by Anton-Latukha** [comment]
> How to switch dynamically the MonadStore instance based on the Options we have (--readonly, --store, etc ?). From what I understand from Haskell, type classes may not be the right tool.

It sounds familiar to what Type Class instances with Type Family data type declaration solve.

I remember that solution by a card game engine story:
The story is:

  "I want to implement an engine for all kinds of card games. I made basic card operations (take, put, play ... etc). I tried to differentiate between card decks and with type classes. But type classes bind to only one data type and that means to the only one type of a deck. Card games have very different decks: Taro decks, 36, 52, decks for RPG games, Uno, 52+Joker."
  
  The guy found that TypeFamilies data type allows supporting all kinds of card decks.
  
Readonly store is for other things then the writable store (& writable store ideally should only write and prevent all other actions), so they can be seen as two types of card games with a store.

**Comment by Anton-Latukha** [comment]
I'd waited until GitHub Actions CI allows to not only show Bool of a workflow, but also mark workflow as a (yellow/orange) warning. Or use some other CI for those checks that we would select as a warning. Otherwise, we would have the implicit dichotomy between the flows and our modes of software development, there are two flows: flow through Hackage and flow through `master` branch. If CI would not be explicit and understandable - the people *may* blame the CI for telling the truth, and the CI maintainer for it. Which is not the case - the people just want to work through two flows at the same time, and without backward compatibility work the projects can support only one of them: or gradual projects development through gradual Hackage releases, OR support `master` state.

If we have "a warning (yellow/orange) level" option for the CI builds - CI maintainers would be able to easily switch between the project build modes. For example, after the release, when the first breaking change arrives in the above projects - the project may start to switch to the `master` and the other build mode would be optional warning-level reports.

BTW, merging the projects means extending the CI builds infrastructure, but that GitHub CI machinery already allows.

Overall I am ready for the merge of repos at any time. Because the HNix would split into ~4 packages, having the 6 in one place seems like no difference.
